amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +153,224 @@ public void stop() {
+            log.error("Incomplete hash range found for stream {} between {} and {}.", streamIdentifier,
+                    mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1));
+            return Optional.of(new HashRangeHole(mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1)));
+        }
+        // Check for any holes in the sorted hashrange intervals.
+        if (mergedHashKeyRanges.size() > 1) {
+            for (int i = 1; i < mergedHashKeyRanges.size(); i++) {
+                final HashKeyRangeForLease hashRangeAtStartOfPossibleHole = mergedHashKeyRanges.get(i - 1);
+                final HashKeyRangeForLease hashRangeAtEndOfPossibleHole = mergedHashKeyRanges.get(i);
+                final BigInteger startOfPossibleHole = hashRangeAtStartOfPossibleHole.endingHashKey();
+                final BigInteger endOfPossibleHole = hashRangeAtEndOfPossibleHole.startingHashKey();
+
+                if (!endOfPossibleHole.subtract(startOfPossibleHole).equals(BigInteger.ONE)) {
+                    log.error("Incomplete hash range found for {} between {} and {}.", streamIdentifier,
+                            hashRangeAtStartOfPossibleHole, hashRangeAtEndOfPossibleHole);
+                    return Optional.of(new HashRangeHole(hashRangeAtStartOfPossibleHole, hashRangeAtEndOfPossibleHole));
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    @VisibleForTesting
+    static List<HashKeyRangeForLease> sortAndMergeOverlappingHashRanges(
+            List<HashKeyRangeForLease> hashKeyRanges) {
+        if (hashKeyRanges.size() == 0 || hashKeyRanges.size() == 1)
+            return hashKeyRanges;
+
+        Collections.sort(hashKeyRanges, new HashKeyRangeComparator());
+
+        final HashKeyRangeForLease first = hashKeyRanges.get(0);
+        BigInteger start = first.startingHashKey();
+        BigInteger end = first.endingHashKey();
+
+        final List<HashKeyRangeForLease> result = new ArrayList<>();
+
+        for (int i = 1; i < hashKeyRanges.size(); i++) {
+            HashKeyRangeForLease current = hashKeyRanges.get(i);
+            if (current.startingHashKey().compareTo(end) <= 0) {
+                end = current.endingHashKey().max(end);
+            } else {
+                result.add(new HashKeyRangeForLease(start, end));
+                start = current.startingHashKey();
+                end = current.endingHashKey();
+            }
+        }
+        result.add(new HashKeyRangeForLease(start, end));
+        return result;
+    }
+
+    @Value
+    private static class HashRangeHole {
+        private final HashKeyRangeForLease hashRangeAtStartOfPossibleHole;
+        private final HashKeyRangeForLease hashRangeAtEndOfPossibleHole;
+    }
+
+    private static class HashRangeHoleTracker {
+        private HashRangeHole hashRangeHole;
```

Comment:
Do we need which hash  range hole  for building state, counter can tell which hash range and also we can emit log for missing hash ranges